src/box/txn.c

```diff
@@ -54,6 +54,170 @@ txn_on_stop(struct trigger *trigger, void *event);
 static int
 txn_on_yield(struct trigger *trigger, void *event);
 
+/** String representation of enum tx_alloc_type. */
+const char *tx_alloc_type_strs[TX_ALLOC_MAX] = {
+	"statements",
+	"user data",
+	"system data",
+};
+
+/** Objects for tx_region_alloc_object method. */
+enum tx_alloc_object {
+	/**
+	 * Object of type struct txn_stmt.
+	 */
+	TX_OBJECT_STMT = 0,
+	/**
+	 * Object of type struct xrow_header.
+	 */
+	TX_OBJECT_XROW_HEADER = 1,
+	/**
+	 * Object of type struct trigger.
+	 */
+	TX_OBJECT_TRIGGER = 2,
+	/**
+	 * Object of type struct ev_timer.
+	 */
+	TX_OBJECT_EV_TIMER = 3,
+	TX_OBJECT_MAX = 4,
+};
+
+/**
+ * Reset txn's alloc_stats.
+ */
+static inline void
+txn_reset_stats(struct txn *txn)
+{
+	memset(txn->alloc_stats, 0, sizeof(txn->alloc_stats));
+}
+
+/**
+ * Collect allocation statistics.
+ */
+static inline void
+tx_track_allocation(struct txn *txn, size_t size, enum tx_alloc_type type)
+{
+	assert(type < TX_ALLOC_MAX);
+	assert(txn != NULL);
+	assert(txn->alloc_stats != NULL);
+	/* Check if txn region is not released. */
+	assert(txn->acquired_region_used == 0);
+
+	txn->alloc_stats[type] += size;
+}
+
+/**
+ * Choose tx_alloc_type for alloc_obj.
+ */
+static inline enum tx_alloc_type
+tx_region_object_to_type(enum tx_alloc_object alloc_obj)
+{
+	enum tx_alloc_type alloc_type = TX_ALLOC_MAX;
+	switch (alloc_obj) {
+	case TX_OBJECT_STMT:
+		alloc_type = TX_ALLOC_STMT;
+		break;
+	case TX_OBJECT_XROW_HEADER:
+	case TX_OBJECT_TRIGGER:
+	case TX_OBJECT_EV_TIMER:
+		alloc_type = TX_ALLOC_SYSTEM;
+		break;
+	default:
+		unreachable();
+		panic("Unexpected alloc_obj in %s", __func__);
+	};
+	assert(alloc_type < TX_ALLOC_MAX);
+	return alloc_type;
+}
+
+/**
+ * Alloc object on region. Pass object as enum tx_alloc_object.
+ * Use this method to track txn's allocations!
+ */
+static inline void *
+tx_region_alloc_object(struct txn *txn, enum tx_alloc_object alloc_obj,
+		       size_t *size)
+{
+	size_t alloc_size = 0;
+	void *alloc = NULL;
+	enum tx_alloc_type alloc_type = tx_region_object_to_type(alloc_obj);
+	switch (alloc_obj) {
+	case TX_OBJECT_STMT:
+		alloc = region_alloc_object(&txn->region,
+					    struct txn_stmt, &alloc_size);
+		break;
+	case TX_OBJECT_TRIGGER:
+		alloc = region_alloc_object(&txn->region,
+					    struct trigger, &alloc_size);
+		break;
+	case TX_OBJECT_XROW_HEADER:
+		alloc = region_alloc_object(&txn->region,
+					    struct xrow_header, &alloc_size);
+		break;
+	case TX_OBJECT_EV_TIMER:
+		alloc = region_alloc_object(&txn->region,
+					    struct ev_timer, &alloc_size);
+		break;
+	default:
+		unreachable();
+		panic("Unexpected alloc_obj in %s", __func__);
```

Comment:
AFAIC, there mustn't be anything after `unreachable`: the program is undefined at this point: https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html.

Reply:
fixed